lib/jsonrpc_client.ml

```diff
@@ -26,6 +26,40 @@ let json_rpc_write_timeout = ref 60000000000L (* timeout value in ns when writin
 
 let to_s  s = (Int64.to_float s) *. 1e-9
 
+let retrieve_json_str str =
+	let buf = Buffer.create 1024 in
+	let fetch_indx = ref 0 in
+	let str_len = String.length str in
+	let input_char () =
+		if !fetch_indx < str_len then
+		begin
+			let chr = String.get str !fetch_indx in
+			fetch_indx := !fetch_indx + 1;
+			chr
+		end
+		else raise Yojson.End_of_input
+	in
+	let brace_cnt = ref (-1) in (* made to tolerant heading space *)
+	let in_string = ref false in
+	let last_char () = Buffer.nth buf (Buffer.length buf - 1) in
+	let rec get () =
+		let c = input_char () in
+		begin
+			match c with
+			| '{' when not !in_string ->
```

Comment:
I think it's a pity if we need to do all this low-level parsing here again (we just deleted a function like this). A JSON library should be doing it. Can we instead modify ocaml-rpc to have an option to filter out "junk" at the end of a JSON-RPC request?

Reply:
Instead of implementing our own JSON parser wouldn't `Yojson.Basic.from_lexbuf ~stream:true` give us enough control to skip everything after the JSON value? (alternatively Jsonm should also give us similar low-level control)

Reply:
I tried, but ocaml-rpc does not expose many of its internals so we cannot simply rewrite the logic there. For Yojson, I made changes like following:
module Yojson_private = struct
  include Yojson.Safe

  let from_lexbuf v ?(stream = false) lexbuf =
      read_space v lexbuf;
      if read_eof lexbuf then
        raise Yojson.End_of_input
      else
        read_json v lexbuf

  let from_string ?buf ?fname ?lnum s =
    try
      let lexbuf = Lexing.from_string s in
      let v = init_lexer ?buf ?fname ?lnum () in
      from_lexbuf v lexbuf
    with End_of_input ->
      Yojson.json_error "Blank input data"
end

It now can just read the first json out and ignore what's left. However, as I stated above, this json cannot be sent back to jsonrpc. It does not accept json to convert to Rpc, from a module perspective. If anyone can suggest any simple way we can override jsonrpc's some internal logic, then I can try.

Reply:
Specifically, if we have functions like json_to_rpc and get_response exposed in interface for jsonrpc, then it could be a more straight forwarding change.

Reply:
Could you add a `?(stream=false)` to from_string and propagata it to `from_lexbuf`?

Reply:
It is similar, could be done. But the point is really not how this get done in Yojson. It is we cannot get the json out of Yojson back to jsonrpc.

Reply:
I think a PR to ocaml-rpc to add a way to do streaming seems the best solution here.

Reply:
Shall we merge this in first, as there is a broken product waiting to be fixed/have work around?
Then we can take time to enhance/improve Yojson or ocaml-rpc and integrate those changes to refine codes here.

Besides, not sure what exactly you mean by streaming. If you mean to have ocaml-rpc to be able to get some json one by one, we may need to also largely change Yojson, coz the current Yojson only deal with complete and correct jsons even in streaming mode. And that also takes time. If you mean just get one rpc/json from the input (and ignores what's left) in ocaml-rpc, will they owner of ocaml-rpc accept such change as it seems incomplete?

Reply:
Anyway, the current change won't be worse than what's before CA-236855, where similar code to parse json works.